pkg/service/auth_context.go

```diff
@@ -47,71 +48,121 @@ func NewAuthContext(parentCtx context.Context, req *auth.CheckRequest, apiConfig
-			authContext.Identity[v] = authObj
-		}); err != nil {
-		return err
+// EvaluateIdentity evaluates given identity configs and retrieves the first valid identity object.
+func (authContext *AuthContext) EvaluateIdentity() error {
+	identityConfigs := authContext.API.IdentityConfigs
+	if len(identityConfigs) <= 0 {
+		return errors.New("no identity configs found")
 	}
 
-	// metadata
-	if err := authContext.evaluateAuthConfigs(authContext.API.MetadataConfigs,
-		func(conf common.AuthConfigEvaluator, authObj interface{}) {
+	successChannel := make(chan bool)
+
+	go func() {
+		defer close(successChannel)
+		authContext.evaluateAuthConfigs(identityConfigs,
+			func(conf common.AuthConfigEvaluator, cancelCtx context.CancelFunc, authObj interface{}, err error) error {
+				if err != nil {
+					authCtxLog.Error(err, "error evaluating identity config", "config", conf)
+					return nil
+				}
+				// Convert from interfaceType to SpecificType
+				v, _ := conf.(*config.IdentityConfig)
+				authCtxLog.Info("Identity", "Config", conf, "AuthObj", authObj)
+				authContext.Identity[v] = authObj
+				cancelCtx()
+				successChannel <- true
+				return nil
+			})
+	}()
+	success := <-successChannel
+	if !success {
+		return errors.New("error evaluating identity configs")
+	}
+	return nil
+}
+
+// EvaluateMetadata checks evaluates the given metadata configs and retrieve possible info.
+func (authContext *AuthContext) EvaluateMetadata() {
+	metadataConfigs := authContext.API.MetadataConfigs
+	if len(metadataConfigs) <= 0 {
+		authCtxLog.Info("No metadata configs found", "Configs", metadataConfigs)
+		return
+	}
+	authContext.evaluateAuthConfigs(metadataConfigs,
+		func(conf common.AuthConfigEvaluator, cancelCtx context.CancelFunc, authObj interface{}, err error) error {
+			if err != nil {
+				authCtxLog.Error(err, "error evaluating metadata config", "config", conf)
+				return nil
+			}
 			// Convert from interfaceType to SpecificType
 			v, _ := conf.(*config.MetadataConfig)
-			authCxtLog.Info("Metadata", "Config", conf, "AuthObj", authObj)
+			authCtxLog.Info("Metadata", "Config", conf, "AuthObj", authObj)
 			authContext.Metadata[v] = authObj
-		}); err != nil {
-		return err
-	}
+			return nil
+		})
+}
 
-	// policy enforcement (authorization)
-	if err := authContext.evaluateAuthConfigs(authContext.API.AuthorizationConfigs,
-		func(conf common.AuthConfigEvaluator, authObj interface{}) {
+// EvaluateAuthorization evaluates the authorization config, along with any metadata and identity and authorizes it.
+func (authContext *AuthContext) EvaluateAuthorization() error {
+	authorizationConfigs := authContext.API.AuthorizationConfigs
+	if len(authorizationConfigs) <= 0 {
+		return errors.New("no authorization configs found")
+	}
+	return authContext.evaluateAuthConfigs(authContext.API.AuthorizationConfigs,
+		func(conf common.AuthConfigEvaluator, cancelCtx context.CancelFunc, authObj interface{}, err error) error {
```

Comment:
Apparently the reason for this function to return an `error` now is only because of the evaluation of the autorization configs, where an error is blocking and should close the conext. Why not making this function a routine that does not return anything and use `cancelCtx` here?